src/librustc_errors/emitter.rs

```diff
@@ -268,132 +269,216 @@ pub trait Emitter {
+                // "in this expansion of" label above is always added in that mode,
+                // and it needs an "in this macro invocation" label to match that.
+                let redundant_span = trace.call_site.contains(sp);
+
+                if !redundant_span &&
                     !trace.macro_decl_name.starts_with("desugaring of ") &&
                     !trace.macro_decl_name.starts_with("#[") ||
                     always_backtrace {
                     new_labels.push((trace.call_site,
                                         format!("in this macro invocation{}",
-                                                if backtrace_len > 2 && always_backtrace {
+                                                if macro_backtrace.len() > 2 && always_backtrace {
                                                     // only specify order when the macro
                                                     // backtrace is multiple levels deep
                                                     format!(" (#{})", i + 1)
                                                 } else {
                                                     String::new()
                                                 })));
-                    if !always_backtrace {
-                        break;
-                    }
+                }
+                if !always_backtrace {
+                    break;
                 }
             }
         }
+
         for (label_span, label_text) in new_labels {
             span.push_span_label(label_span, label_text);
         }
-        for sp_label in span.span_labels() {
-            if sp_label.span.is_dummy() {
-                continue;
-            }
-            if sm.span_to_filename(sp_label.span.clone()).is_macros() &&
-                !always_backtrace
-            {
-                let v = sp_label.span.macro_backtrace();
-                if let Some(use_site) = v.last() {
-                    before_after.push((sp_label.span.clone(), use_site.call_site.clone()));
-                }
-            }
+    }
+
+    // This does a small "fix" for multispans by looking to see if it can find any that
+    // point directly at <*macros>. Since these are often difficult to read, this
+    // will change the span to point at the use site.
+    fn fix_multispans_in_extern_macros(
+        &self,
+        source_map: &Option<Lrc<SourceMap>>,
+        span: &mut MultiSpan,
+        children: &mut Vec<SubDiagnostic>,
+    ) {
+        for span in iter::once(span).chain(children.iter_mut().map(|child| &mut child.span)) {
+            self.fix_multispan_in_extern_macros(source_map, span);
         }
+    }
+
+    // This "fixes" MultiSpans that contain Spans that are pointing to locations inside of
+    // <*macros>. Since these locations are often difficult to read, we move these Spans from
+    // <*macros> to their corresponding use site.
+    fn fix_multispan_in_extern_macros(
```

Comment:
```suggestion
    fn fix_multispan_in_macros(
```

Reply:
The `is_macros` method on `FileName` should really be named `is_extern_macro`.
That `FileName` categorization doesn't go away with this PR, only after #66364 (and even then, we might want to keep cross-crate macro spans out of diagnostics, even if `<... macros>` would be gone).

Reply:
Then I don't understand why the option was renamed from `-Zexternal-macro-backtrace` to `-Zmacro-backtrace`.

The only difference between this option enabled and not enabled (besides the "this error originates in a macro" message) is `fix_multispans_in_extern_macros`, which affects only extern macros.

Reply:
Also, why is the "this error originates in a macro" message recommends enabling `-Z macro-backtrace` for non-extern macros where it makes no difference?

Reply:
What `-Z macro-backtrace` does (and `-Z external-macro-backtrace` did, before this PR), is control whether the full backtrace is displayed, *regardless* of whether the macros are from.

Most of the difference is made by the `if !always_backtrace { break; }` in `render_multispan_macro_backtrace`, which means that not providing the flag makes it always stop after one frame.

The only reason `fix_multispans_in_extern_macros` is skipped when the flag is provided, is because it would destroy the information needed to show a full backtrace.

Reply:
>Most of the difference is made by the if !always_backtrace { break; } in render_multispan_macro_backtrace, which means that not providing the flag makes it always stop after one frame.

Ok, I finally see the difference.

Could you add a test with an error in local macros and two revisions - with and without `-Z macro-backtrace`?
(Or it can be added to the existing test `ui\macro_backtrace\main.rs`.)